### PR TITLE
feat(storage): bump opendal to v0.49.0

### PR DESCRIPTION
**Pull request by chenzl25**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- As title

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by wcy-fdu** [comment]
Is there any special reason to bump? I think every time we bump OpenDAL we need to do a lot of testing in aws, gcp and azure environments.

**Comment by chenzl25** [comment]
> Is there any special reason to bump? I think every time we bump OpenDAL we need to do a lot of testing in aws, gcp and azure environments.

No special purpose 🥵 . Previously I suspected some bug in the low-level opendal, but actually not. However, I already have finished the coding part, so just create a PR by the way.

**Comment by wcy-fdu** [comment]
> > Is there any special reason to bump? I think every time we bump OpenDAL we need to do a lot of testing in aws, gcp and azure environments.
> 
> No special purpose 🥵 . Previously I suspected some bug in the low-level opendal, but actually not. However, I already have finished the coding part, so just create a PR by the way.

Thanks for the efforts first🙌. 
As we already do some test on OpenDAL 0.47 for realsed 2.0, do you think we should bump to 0.49 as well? Or just hold this pr for a while and merge it when needed. I have no preference, I can help with testing if needed. It's up to you to decide cc @hzxa21

**Comment by hzxa21** [comment]
> > > Is there any special reason to bump? I think every time we bump OpenDAL we need to do a lot of testing in aws, gcp and azure environments.
> > 
> > 
> > No special purpose 🥵 . Previously I suspected some bug in the low-level opendal, but actually not. However, I already have finished the coding part, so just create a PR by the way.
> 
> Thanks for the efforts first🙌. As we already do some test on OpenDAL 0.47 for realsed 2.0, Patrick do you think we should bump to 0.49 as well? Or just hold this pr for a while and merge it when needed. I have no preference, I can help with testing if needed. It's up to you to decide cc @hzxa21

There is no technical blocker of upgrading OpenDAL to the latest version if we think the latest version is also stable. We just need to run all the tests needed to make sure we are confident about this version before merging.

To reduce risks, I suggest we don't bump OpenDAL to latest in 2.0 given that 0.47 is the version we have proved working with several rounds of bug fixing and testing. I think we can certify OpenDAL 0.49 with the same tests we did for 0.47 before merging this PR to main, without blocking release 2.0. @wcy-fdu can decide the schedule because you are more familiar with OpenDAL than us.

**Comment by wcy-fdu** [comment]
> can decide the schedule because you are more familiar with OpenDAL than us.

Let's first run all the tests based on this pr, and merge this pr into main if all pass.
And let’s **not** cherry pick version 2.0 yet. We can continue testing during the release test of version 2.1.

I will run test based on this PR later, so just hold for a while.

File: src/object_store/src/object/opendal_engine/opendal_object_store.rs
Comment (BugenZhao) [inline src/object_store/src/object/opendal_engine/opendal_object_store.rs:393]:
```suggestion
                Err(e) => tracing::info!(
                    path = ?self.path,
                    error = %e.as_report(),
                    "File can not be stated, and fail when writer write.",
                ),
```

Can refer to https://www.notion.so/risingwave-labs/A-Guide-to-Error-Handling-that-Just-Works-Part-I-e108f8d0845b4bddaa0e5843791a8cac?pvs=4#f781cf9ecaee436a97adf9f051f2ce99

**Review by wcy-fdu — APPROVED** [review]
I run longevity test and perf test based on this branch, and both look good to me. So I think this upgrade is safe. 
I will resolve the conflict.
cc @hzxa21 for another review.

File: Cargo.lock
Comment (hzxa21) [inline Cargo.lock:7892]: Shall we upgrade to 0.49.2 instead of 0.49.0?

**Review by hzxa21 — APPROVED** [review]
LGTM

File: Cargo.lock
Comment (wcy-fdu) [reply Cargo.lock:7892]: OpenDAL just release v0.50 5days ago🥲.

Comment (hzxa21) [reply Cargo.lock:7892]: We don't need to rush to 0.50. Given that we specify `0.49` in `Carago.toml`, I think we should bump to the latest `0.49` which is `0.49.2`

**Pull request merged**
